File: fx/src/main/java/gargoyle/fx/FXCloseFlag.java
package gargoyle.fx;

import javafx.collections.ObservableMap;
import javafx.util.Callback;

enum FXCloseFlag {
    RESTART,
    PREVENT;

    @SuppressWarnings("unchecked")
    FXCloseAction doIf(final FXContext context) {
        final ObservableMap<Object, Object> properties = context.getStage().getProperties();
        if (properties.containsKey(this)) {
            final Callback<FXContext, FXCloseAction> callback = (Callback<FXContext, FXCloseAction>)
                    properties.get(this);
            properties.remove(this);
            return callback.call(context);
        }
        return FXCloseAction.CLOSE;
    }

    void set(final FXContext context, final Callback<FXContext, FXCloseAction> callback) {
        context.getStage().getProperties().put(this, callback);
    }
}

File: fx/src/main/kotlin/gargoyle/fx/FxCloseFlag.kt
package gargoyle.fx

internal enum class FxCloseFlag {
    RESTART, PREVENT;

    @Suppress("UNCHECKED_CAST")
    fun doIf(context: FxContext): FxCloseAction =
        context.stage.properties.let { properties ->
            if (properties.containsKey(this)) {
                (properties[this] as (FxContext) -> FxCloseAction)(context).also {
                    properties.remove(this)
                }
            } else
                FxCloseAction.CLOSE
        }

    operator fun set(context: FxContext, callback: (FxContext) -> FxCloseAction) {
        context.stage.properties[this] = callback
    }
}
